Sources/CartonCLI/Commands/Init.swift

```
@@ -25,43 +25,85 @@ struct Init: ParsableCommand {
+    } catch let nsError as NSError {
+        throw ProjectInitialisationError.cannotCreateDirectory(path: packagePathRelativeToHomeDirectory, reason: nsError.localizedDescription)
+    }
+    
+    // FIXME: Handle this error-catching in each `Template.create()` instead.
+    do {
+      // FIXME:  The first `template` is of type `Templates`, and two consecutive `template` looks odd.
+      // Maybe rename `Templates` as `TemplateKind`, and refactor `Template` so the template doesn't have to be created this way.
+      try template.template.createProject(
+        at: projectPath,
+        inSubdirectory: name != nil,
+        on: localFileSystem,
+        terminal: terminal
+      )
+    } catch let nsError as NSError {
+        throw ProjectInitialisationError.cannotCreateWithTemplate(template, reason: nsError.localizedDescription)
+    } catch let error {
+        throw ProjectInitialisationError.cannotCreateWithTemplate(template, reason: "\(error)")
     }
-    try localFileSystem.createDirectory(packagePath)
-    try template.template.create(
-      on: localFileSystem,
-      project: .init(name: name, path: packagePath, inPlace: self.name == nil),
-      terminal
-    )
   }
 }
+
+// TODO: Add a diagnostics engine for errors thrown.
+
+enum ProjectInitialisationError: Error, CustomStringConvertible {
+    case cannotFindCurrentDirectory
+    case cannotCreateDirectory(path: RelativePath, reason: String)
+    case cannotCreateWithTemplate(Templates, reason: String)
+    
+    var description: String {
+        switch self {
+        case .cannotFindCurrentDirectory:
+            return "Cannot find the current working directory."
+        case let .cannotCreateDirectory(path, reason):
+            return "Cannot create directory ~/\(path.pathString): \(reason)"
+        case let .cannotCreateWithTemplate(template, reason):
+            return "Cannot create project with template '\(template)': \(reason)"
+        }
+    }
+}
```

It seems more appropriate to me that carton should throw an error instead of simply writing a message on the terminal. But we probably need a diagnostics engine to fully take advantage of it.

---

Yes, the reason for writing terminal messages is that thrown errors aren't nicely formatted or highlighted by default.